File: maxSumAfterPartitioning.java
class Solution {
    public int maxSumAfterPartitioning(int[] arr, int k) {
        
        int n=arr.length;
        int[] dp=new int[n];
        
        int max=-1;
        
        for(int i=0;i<k;i++)
        {
            max=Math.max(max,arr[i]);
            dp[i]=max*(i+1);
        }
        
        max=-1;
        for(int i=k;i<n;i++)
        {
            max=arr[i];
            for(int j=0;j<k;j++)
            {
                max=Math.max(max,arr[i-j]);
                dp[i]=Math.max(dp[i],dp[i-j-1]+max*(j+1));
                
            }
        }
        
        return dp[n-1];
        
    }
}

// Time Complexity : O(N*K)
// Space Complexity : O(N)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no

File: maxSumAfterPartitioning.cpp
// Time Complexity : O(n * k) 
// Space Complexity : O(n)
//          where n :- size of i/p array
// Did this code successfully run on Leetcode : yes

// Three line explanation of solution in plain english
/* Build DP array to store maximum partition value that can be achieved till current index.
 * Iterate over each element, while iterating over each element go back till k elements (k choices) to compute 
 * the max possible value.
 * Use DP array to get previous partion values which are already computed. 
 */

class Solution {
public:
    int maxSumAfterPartitioning(vector<int>& arr, int k) {
        if (arr.size() == 0)
        {
            return 0;
        }
        
        int dp[arr.size()];
        
        for (int i = 0; i < arr.size(); i++)
        {
            dp[i] = 0;
        }
        
        dp[0] = arr[0];
        
        for (int i = 1; i < arr.size(); i++)
        {
            int max = arr[i];
            
            for (int j = 1; j <= k && i - j + 1 >= 0; j++)
            {
                max = std::max(max, arr[i - j + 1]);
                
                if (i - j >= 0)
                {
                    dp[i] = std::max(dp[i], dp[i - j] + max * j);
                }
                else
                {
                    dp[i] = std::max(dp[i], max * j);
                }
            }
        }
        
        return dp[arr.size() - 1];
    }
};